server.go

```diff
@@ -456,7 +458,91 @@ func (s *Server) handle(ic ircConn) {
 	s.metrics.downstreams.Add(-1)
 }
 
-func (s *Server) Serve(ln net.Listener) error {
+func (s *Server) HandleAdmin(ic ircConn) {
+	defer func() {
+		if err := recover(); err != nil {
+			s.Logger.Printf("panic serving admin client %q: %v\n%v", ic.RemoteAddr(), err, string(debug.Stack()))
+		}
+	}()
+
+	s.lock.Lock()
+	shutdown := s.shutdown
+	s.lock.Unlock()
+
+	ctx := context.TODO()
+	remoteAddr := ic.RemoteAddr().String()
+	logger := &prefixLogger{s.Logger, fmt.Sprintf("admin %q: ", remoteAddr)}
+	c := newConn(s, ic, &connOptions{Logger: logger})
+
+	if shutdown {
+		c.SendMessage(ctx, &irc.Message{
+			Command: "ERROR",
+			Params:  []string{"Server is shutting down"},
+		})
+	} else {
+		for {
+			msg, err := c.ReadMessage()
+			if errors.Is(err, io.EOF) {
+				break
+			} else if err != nil {
+				logger.Printf("failed to read IRC command: %v", err)
+				break
+			}
+			switch msg.Command {
+			case "BOUNCERSERV":
+				if len(msg.Params) < 1 {
+					c.SendMessage(ctx, &irc.Message{
+						Command: irc.ERR_NEEDMOREPARAMS,
+						Params: []string{
+							"*",
+							msg.Command,
+							"Not enough parameters",
+						},
+					})
+					break
+				}
+				err := handleServicePRIVMSG(&serviceContext{
+					Context: ctx,
+					srv:     s,
+					admin:   true,
+					print: func(text string) {
+						c.SendMessage(ctx, &irc.Message{
+							Prefix:  s.prefix(),
+							Command: "PRIVMSG",
+							Params:  []string{"*", text},
+						})
+					},
+				}, msg.Params[0])
+				if err != nil {
+					c.SendMessage(ctx, &irc.Message{
+						Prefix:  s.prefix(),
+						Command: "FAIL",
+						Params:  []string{msg.Command, err.Error()},
+					})
+				} else {
+					c.SendMessage(ctx, &irc.Message{
+						Prefix:  s.prefix(),
+						Command: msg.Command,
+						Params:  []string{"OK"},
+					})
+				}
+			default:
+				c.SendMessage(ctx, &irc.Message{
+					Prefix:  s.prefix(),
+					Command: irc.ERR_UNKNOWNCOMMAND,
+					Params: []string{
+						"*",
+						msg.Command,
+						"Unknown command",
+					},
+				})
+			}
+		}
+	}
+	c.Close()
+}
+
+func (s *Server) Serve(ln net.Listener, handler func(ircConn)) error {
```

Comment:
I'm not a super fan of making this a generic server framework… Another way to go about this would be to add an `admin bool` flag instead for instance (hm, OTOH would make it less explicit that passing true has security implications…). But I don't mind too much.

Reply:
🙅‍♂️